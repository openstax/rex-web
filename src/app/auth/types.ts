export interface User {
  firstName: string;
  lastName: string;
  uuid: string;
  isNotGdprLocation: boolean;
}

export interface State {
  user: User | undefined;
  established: boolean;
}
