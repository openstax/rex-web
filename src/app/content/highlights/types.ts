import { Highlight, HighlightColorEnum, HighlightsSummary, NewHighlight } from '@openstax/highlighter/dist/api';
import { ScrollTarget } from '../../navigation/types';
import { FiltersChange } from '../components/popUp/types';
import { LinkedArchiveTreeNode } from '../types';

export interface SummaryHighlights {
  [locationId: string]: {[pageId: string]: HighlightData[]};
}

export type OrderedSummaryHighlights = Array<{
  location: LinkedArchiveTreeNode,
  pages: Array<{
    pageId: string;
    highlights: HighlightData[];
  }>
}>;

export interface SummaryFilters {
  locationIds: string[];
  colors: HighlightColorEnum[];
}

export interface SummaryFiltersUpdate {
  locations?: FiltersChange<LinkedArchiveTreeNode>;
  colors?: FiltersChange<HighlightColorEnum>;
}

export type CountsPerSource = NonNullable<HighlightsSummary['countsPerSource']>;
export type HighlightColorCounts = CountsPerSource[string];

export type SummaryHighlightsPagination = null | {
  sourceIds: string[];
  page: number;
  perPage: number;
};

export interface State {
  currentPage: {
    pageId: string | null,
    highlights: null | HighlightData[];
    hasUnsavedHighlight: boolean;
    focused?: string;
  };
  summary: {
    open: boolean,
    pagination: SummaryHighlightsPagination,
    totalCountsPerPage: CountsPerSource | null;
    filters: SummaryFilters,
    loading: boolean;
    highlights: SummaryHighlights | null;
  };
}

export interface HighlightScrollTarget extends ScrollTarget {
  type: 'highlight';
  id: string;
}

// scopeId should not be nullable in the swagger
export type NewHighlightPayload = NewHighlight & {scopeId: string};
export type HighlightData = Highlight & {scopeId: string};
