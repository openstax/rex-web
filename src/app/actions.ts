import { createStandardAction } from 'typesafe-actions';

export const receivePageFocus = createStandardAction('app/receivePageFocus')<boolean>();
