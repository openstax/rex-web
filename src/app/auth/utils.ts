import { AccountsUser } from '../../gateways/createUserLoader';

export const formatUser = (user: AccountsUser) => ({
  firstName: user.first_name,
  isNotGdprLocation: user.is_not_gdpr_location,
  lastName: user.last_name,
  uuid: user.uuid,
});
