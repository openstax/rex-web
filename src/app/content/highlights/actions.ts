import { Highlight, NewHighlight, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { LocationFilters } from '../components/popUp/types';
import {
  CountsPerSource,
  HighlightData,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from './types';

export const focusHighlight = createStandardAction('Content/Highlights/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlights/clear')();
export const createHighlight = createStandardAction('Content/Highlights/create')<NewHighlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const requestDeleteHighlight = createStandardAction('Content/Highlights/requestDelete')<Highlight, {
  locationFilterId: string,
  pageId: string,
}>();
export const receiveDeleteHighlight = createStandardAction('Content/Highlights/receiveDelete')<Highlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const updateHighlight = createStandardAction('Content/Highlights/update')<UpdateHighlightRequest, {
  revertingAfterFailure?: boolean,
  preUpdateData: UpdateHighlightRequest,
  locationFilterId: string,
  pageId: string,
}>();
export const receiveHighlights = createStandardAction(
  'Content/Highlights/receive'
)<{highlights: HighlightData[], pageId: string}>();
export const setAnnotationChangesPending = createStandardAction('Content/Highlights/setAnnotationChangesPending')<
  boolean
>();

export const openMyHighlights = createStandardAction('Content/Highlights/Summary/open')<void>();
export const closeMyHighlights = createStandardAction('Content/Highlights/Summary/close')<void>();
export const initializeMyHighlightsSummary = createStandardAction('Content/Highlights/Summary/init')<void>();

export const printSummaryHighlights = createStandardAction('Content/Highlights/Summary/print')();
export const toggleSummaryHighlightsLoading = createStandardAction('Content/Highlights/Summary/loading')<boolean>();

export const loadMoreSummaryHighlights = createStandardAction('Content/Highlights/Summary/loadMore')();
export const setSummaryFilters = createStandardAction('Content/Highlights/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/Highlights/Summary/updateFilters')<
  Partial<SummaryFiltersUpdate>
>();
export const receiveSummaryHighlights = createStandardAction('Content/Highlights/Summary/receiveHighlights')<
  SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: SummaryFilters,
    isStillLoading?: boolean
  }
>();
export const receiveHighlightsTotalCounts = createStandardAction(
  'Content/receiveHighlightsTotalCounts'
)<CountsPerSource, LocationFilters>();
