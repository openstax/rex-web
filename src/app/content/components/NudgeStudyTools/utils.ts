import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize, useOnScrollTopOffset } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { Page } from '../../types';
import {
  arrowDesktopHeight,
  arrowRightMargin,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  closeButtonSize,
  contentMarginLeft,
  contentMarginTop,
  contentWidth,
  cookieNudge,
  daysUntilCookieExpires,
  mobileNudgeStudyToolsTargetId,
  mobileSpotlightPadding,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperLeft: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean): Positions => {
  const padding = remsToPx(mobileSpotlightPadding);
  const { top, left, height, width } = getBoundingRectOfNudgeTarget(target);
  const spotlightLeftOffset = left - (isMobile ? padding : 0);
  const spotlightWidth = width + (isMobile ? padding * 2 : 0);
  const arrowTopOffset = top + height + remsToPx(arrowTopMargin);
  // left edge of arrow image should be on the middle of the spotlight (adjusted for a margin)
  const arrowLeft = spotlightLeftOffset + (spotlightWidth / 2) - remsToPx(arrowRightMargin);

  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperLeft = spotlightLeftOffset + spotlightWidth + remsToPx(contentMarginLeft);
  const closeButtonLeft = contentWrapperLeft + remsToPx(contentWidth) + remsToPx(closeButtonDistanceFromContent);
  const closeButtonTopOffset = contentWrapperTopOffset
    - remsToPx(closeButtonSize)
    - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperLeft,
    contentWrapperTopOffset,
    spotlightHeight: height,
    spotlightLeftOffset,
    spotlightTopOffset: top,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  const { top, bottom, left, right } = Array.from(target.children).reduce((rects, child) => {
    const rect = child.getBoundingClientRect();
    rects.top = Math.min(rects.top, rect.top);
    rects.bottom = Math.max(rects.bottom, rect.bottom);
    rects.left = Math.min(rects.left, rect.left);
    rects.right = Math.max(rects.right, rect.right);
    return rects;
    // Starting values depends on if we checks for Math.max or Math.min
  }, { top: 9999, bottom: 0, left: 9999, right: 0 });

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const scrollTopOffset = useOnScrollTopOffset();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(
      `#${isMobile ? mobileNudgeStudyToolsTargetId : nudgeStudyToolsTargetId}`
    ) as HTMLElement | null;
    if (target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results in different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [scrollTopOffset, studyGuides, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  const newValue = counter + 1;
  Cookies.set(cookieNudge.pageCounter, newValue.toString(), {expires: daysUntilCookieExpires});
  return newValue;
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString(), {expires: daysUntilCookieExpires});
  Cookies.set(cookieNudge.date, now.toString(), {expires: daysUntilCookieExpires});
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  const [counter, updateCounter] = useState<number>(getPageCounterCookie());
  const pageRef = React.useRef<Page>();

  React.useEffect(() => {
    if (page && page !== pageRef.current && counter <= nudgeStudyToolsMinPageLimit) {
      pageRef.current = page;
      updateCounter(incrementPageCounterCookie());
    }
  }, [page, updateCounter, counter]);

  return counter;
};
