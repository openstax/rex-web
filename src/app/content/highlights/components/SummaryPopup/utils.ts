import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
