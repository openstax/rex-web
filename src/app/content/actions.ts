import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
