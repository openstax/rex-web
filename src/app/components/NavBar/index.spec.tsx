import ReactTestUtils from 'react-dom/test-utils';
import createTestStore from '../../../test/createTestStore';
import { reactAndFriends, resetModules } from '../../../test/utils';
import { receiveLoggedOut, receiveUser } from '../../auth/actions';
import { User } from '../../auth/types';
import { Store } from '../../types';
import { assertWindow } from '../../utils';
import { assertNotNull } from '../../utils/assertions';

describe('content', () => {
  let React: ReturnType<typeof reactAndFriends>['React']; // tslint:disable-line:variable-name
  let renderer: ReturnType<typeof reactAndFriends>['renderer'];
  let renderToDom: ReturnType<typeof reactAndFriends>['renderToDom'];
  let TestContainer: ReturnType<typeof reactAndFriends>['TestContainer']; // tslint:disable-line:variable-name

  beforeEach(() => {
    resetModules();
    jest.resetAllMocks();
    ({React, renderer, renderToDom, TestContainer} = reactAndFriends());
  });

  describe('in browser', () => {
    // tslint:disable-next-line:variable-name
    let NavBar: any;
    // tslint:disable-next-line:variable-name
    let store: Store;
    let user: User;

    beforeEach(() => {
      user = {firstName: 'test', isNotGdprLocation: true, uuid: 'some_uuid'};
      store = createTestStore();
      NavBar = require('.').default;
    });

    const render = () => <TestContainer store={store}>
      <NavBar />
    </TestContainer>;

    it('matches snapshot for null state', () => {
      const component = renderer.create(render());
      const tree = component.toJSON();
      component.unmount();

      expect(tree).toMatchSnapshot();
    });

    describe('when logged in', () => {
      beforeEach(() => {
        store.dispatch(receiveUser(user));
      });

      it('matches snapshot', () => {
        const component = renderer.create(render());
        const tree = component.toJSON();
        component.unmount();

        expect(tree).toMatchSnapshot();
      });

      it('doesn\'t prevent default on accounts links', async() => {
        const window = assertWindow();
        const {root} = renderToDom(render());
        const link1 = root.querySelector('a[href^="/accounts/logout"]');
        const link2 = root.querySelector('a[href="/accounts/profile"]');

        if (!link1 || !link2) {
          expect(link1).toBeTruthy();
          expect(link2).toBeTruthy();
          return;
        }

        const event = window.document.createEvent('MouseEvents');
        event.initEvent('click', true, true);
        const preventDefault = event.preventDefault = jest.fn();

        link1.dispatchEvent(event); // this checks for bindings using addEventListener
        ReactTestUtils.Simulate.click(link1, {preventDefault}); // this checks for react onClick prop
        link2.dispatchEvent(event);
        ReactTestUtils.Simulate.click(link2, {preventDefault});

        expect(event.preventDefault).not.toHaveBeenCalled();
      });
    });

    it('matches snapshot for logged out', () => {
      store.dispatch(receiveLoggedOut());

      const component = renderer.create(render());
      const tree = component.toJSON();
      component.unmount();

      expect(tree).toMatchSnapshot();
    });

    describe('manages scroll based on ovelay', () => {
      let window: Window;
      let getComputedStyle: jest.SpyInstance;
      let getComputedStyleBack: Window['getComputedStyle'];

      beforeEach(() => {
        store.dispatch(receiveUser(user));
        window = assertWindow();
        getComputedStyleBack = window.getComputedStyle;
        getComputedStyle = window.getComputedStyle = jest.fn();
      });

      afterEach(() => {
        window.getComputedStyle = getComputedStyleBack;
      });

      it('blocks scroll when shown', () => {
        const {node} = renderToDom(render());
        const overlay = assertNotNull(node.querySelector('[data-testid=\'nav-overlay\']'), '');

        getComputedStyle.mockReturnValue({height: '10px'});

        const event = window.document.createEvent('UIEvents');
        event.initEvent('scroll', true, false);
        const preventDefault = jest.spyOn(event, 'preventDefault');

        window.document.dispatchEvent(event);

        expect(getComputedStyle).toHaveBeenCalledWith(overlay);
        expect(preventDefault).toHaveBeenCalled();
      });

      it('allows scroll when hidden', () => {
        const {node} = renderToDom(render());
        const overlay = assertNotNull(node.querySelector('[data-testid=\'nav-overlay\']'), '');

        getComputedStyle.mockReturnValue({height: '0px'});

        const event = window.document.createEvent('UIEvents');
        event.initEvent('scroll', true, false);
        const preventDefault = jest.spyOn(event, 'preventDefault');

        window.document.dispatchEvent(event);

        expect(getComputedStyle).toHaveBeenCalledWith(overlay);
        expect(preventDefault).not.toHaveBeenCalled();
      });

      it('noops without a dom', () => {
        const element = renderer.create(render());
        const OnScroll = require('../OnScroll').default; // tslint:disable-line:variable-name
        const onScroll = element.root.findByType(OnScroll);

        const event = window.document.createEvent('UIEvents');
        event.initEvent('scroll', true, false);
        const preventDefault = jest.spyOn(event, 'preventDefault');

        onScroll.props.callback(event);

        expect(getComputedStyle).not.toHaveBeenCalled();
        expect(preventDefault).not.toHaveBeenCalled();
      });
    });
  });

  describe('polyfill', () => {
    let loaded: boolean;

    beforeEach(() => {
      loaded = false;

      jest.doMock('focus-within-polyfill', () => {
        loaded = true;
      });
    });

    describe('inside browser', () => {
      it('loads', async() => {
        require('.');
        await Promise.resolve();
        expect(loaded).toBe(true);
      });
    });

    describe('outside of browser', () => {
      const documentBack = document;
      const windowBack = window;

      beforeEach(() => {
        delete (global as any).document;
        delete (global as any).window;
      });

      afterEach(() => {
        (global as any).document = documentBack;
        (global as any).window = windowBack;
      });

      it('doesn\'t load', async() => {
        require('.');
        await Promise.resolve();
        expect(loaded).toBe(false);
      });
    });
  });
});
