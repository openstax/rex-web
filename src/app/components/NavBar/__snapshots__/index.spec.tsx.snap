// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content in browser matches snapshot for logged out 1`] = `
<div
  className="styled__BarWrapper-sc-3syvnw-12 ePgITh"
  data-analytics-region="openstax-navbar"
>
  <div
    className="styled__TopBar-sc-3syvnw-0 iVnlSK"
    data-testid="navbar"
  >
    <a
      href="/"
    >
      <img
        alt="OpenStax Logo"
        className="styled__HeaderImage-sc-3syvnw-1 gTGoMI"
        role="img"
        src="logo.svg"
      />
    </a>
    <a
      className="styled__Link-sc-3syvnw-6 bqAshF"
      data-analytics-label="login"
      data-testid="nav-login"
      href="/accounts/login?r=/"
    >
       
      Log in
    </a>
  </div>
</div>
`;

exports[`content in browser matches snapshot for null state 1`] = `
<div
  className="styled__BarWrapper-sc-3syvnw-12 ePgITh"
  data-analytics-region="openstax-navbar"
>
  <div
    className="styled__TopBar-sc-3syvnw-0 iVnlSK"
    data-testid="navbar"
  >
    <a
      href="/"
    >
      <img
        alt="OpenStax Logo"
        className="styled__HeaderImage-sc-3syvnw-1 gTGoMI"
        role="img"
        src="logo.svg"
      />
    </a>
  </div>
</div>
`;

exports[`content in browser when logged in matches snapshot 1`] = `
<div
  className="styled__BarWrapper-sc-3syvnw-12 ePgITh"
  data-analytics-region="openstax-navbar"
>
  <div
    className="styled__TopBar-sc-3syvnw-0 iVnlSK"
    data-testid="navbar"
  >
    <a
      href="/"
    >
      <img
        alt="OpenStax Logo"
        className="styled__HeaderImage-sc-3syvnw-1 gTGoMI"
        role="img"
        src="logo.svg"
      />
    </a>
    <div
      className="styled__DropdownContainer-sc-3syvnw-2 cgIyER"
      data-testid="user-nav"
    >
      <span
        className="styled__DropdownToggle-sc-3syvnw-7 bCCQDs"
        data-testid="user-nav-toggle"
        tabIndex="0"
      >
        Hi test
        <svg
          aria-hidden="true"
          className="styled__DownIcon-sc-3syvnw-4 kElTll StyledIconBase-sc-bdy9j4 cCvWXa"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 448 512"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
          />
        </svg>
        <svg
          aria-hidden="true"
          className="styled__HamburgerIcon-sc-3syvnw-5 jKOrlZ StyledIconBase-sc-bdy9j4 cCvWXa"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 448 512"
        >
          <path
            d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        className="styled__DropdownOverlay-sc-3syvnw-10 heRIAm"
        data-testid="nav-overlay"
        tabIndex="-1"
      >
        <a
          aria-hidden="true"
          href="/"
          tabIndex={-1}
        >
          <img
            alt="OpenStax Logo"
            className="styled__OverlayLogo-sc-3syvnw-3 dlofMy"
            src="logo.svg"
          />
        </a>
        <div>
          <h4
            className="styled__OverlayHeading-sc-3syvnw-8 keFAHn"
          >
            Hi test
          </h4>
          <ul
            className="styled__DropdownList-sc-3syvnw-9 gKoUbp"
          >
            <li>
              <a
                href="/accounts/profile"
                target="_blank"
              >
                Account Profile
              </a>
            </li>
            <li>
              <a
                href="/accounts/logout?r=/"
              >
                Log out
              </a>
            </li>
          </ul>
        </div>
      </div>
      <button
        aria-hidden="true"
        className="styled__TimesIcon-sc-3syvnw-11 iDjoNu"
        tabIndex={-1}
      >
        <svg
          height="15px"
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </button>
    </div>
  </div>
</div>
`;
